enhancements/ingress/header-case-adjustment.md

```diff
@@ -0,0 +1,335 @@
+
+```
+global
+  # ...
+  h1-case-adjust host Host
+  h1-case-adjust x-forwarded-for X-Forwarded-For
+  # ...
+
+defaults
+  # ...
+  option h1-case-adjust-bogus-client
+  # ...
+```
+
+When a Route has the `haproxy.router.openshift.io/h1-adjust-case=true`
+annotation, the router specifies the `h1-case-adjust-bogus-server` setting on
+the configuration for that Route:
+
+```
+backend ...
+  option h1-case-adjust-bogus-server
+```
+
+The ingress operator translates the value of the new
+`spec.httpHeaders.headerNameCaseAdjustments` API field into the format for the
+`ROUTER_H1_CASE_ADJUST` environment variable and sets this environment variable
+to that value in the router Deployment.
+
+### Risks and Mitigations
+
+OpenShift cannot infer which HTTP header names legacy clients and servers expect
+to be capitalized a certain way, so cluster administrators and application
+developers need to be aware of the issue and configure IngressControllers and
+Routes using this new configuration when updating to OpenShift 4.7.  A release
+note may mitigate this concern; for example:
+
+"OpenShift 4.7 updates to HAProxy 2.2, which down-cases HTTP header names by
+default (for example, `Host: xyz.com` is transformed to `host: xyz.com`).  If
+legacy applications are sensitive to the capitalization of HTTP header names,
+see the IngressController's new `spec.httpHeaders.headerNameCaseAdjustments` API
+field for a solution to accommodate these legacy applications until they can be
+fixed."
+
+## Design Details
+
+### Test Plan
+
+The controller that manages the router Deployment and related resources has unit
+test coverage; for this enhancement, the unit tests have been expanded to cover
+the additional functionality.
+
+The operator has end-to-end tests; for this enhancement, a test has been added
+that (1) creates an IngressController configured with case adjustments, (2)
+creates a Pod and Route for an HTTP application that echoes back requests, (3)
+sends a request to the Route, and (4) verifies that the application echoes back
+the expected headers.
+
+### Graduation Criteria
+
+N/A.
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, the default configuration does not perform any case adjustments.  On
+downgrade, the operator ignores the `spec.httpHeaders.headerNameCaseAdjustments`
+API field, and the HAProxy configuration template ignores unrecognized
+environment variables and Route annotations.
+
+### Version Skew Strategy
+
+N/A.
+
+## Implementation History
+
+## Alternatives
+
+As an alternative to adding this API, we can add a release note telling cluster
+administrators about the issue and suggesting fixing legacy applications to
+follow the HTTP protocol specification.  Note that OpenShift 4.4, 4.5, and 4.6
+do not exhibit the problem unless HTTP/2 is enabled, which arguably already
+gives users ample time to fix broken applications before upgrading to OpenShift
+4.7.
+
+Instead of introducing this API in OpenShift 4.7, we could revert OpenShift 4.7
```

Comment:
This alternative is now the chosen approach, right?

Reply:
Right.  I'll remove that paragraph and revise the preceding one.